File: examples/go/order-domain-event/order/service_test.go
package order

import "testing"

func TestPlaceOrderPublishesEventToAllHandlers(t *testing.T) {
	bus := NewEventBus()
	notifications := &NotificationHandler{}
	audit := &AuditHandler{}

	bus.Subscribe(notifications.HandleOrderPlaced)
	bus.Subscribe(audit.HandleOrderPlaced)

	service := NewPlacementService(bus)

	placedOrder, err := service.PlaceOrder("ord-5001", "cust-100", 15900)
	if err != nil {
		t.Fatalf("place order: %v", err)
	}

	if placedOrder.ID != "ord-5001" {
		t.Fatalf("expected order id ord-5001, got %s", placedOrder.ID)
	}

	if len(notifications.Messages) != 1 {
		t.Fatalf("expected 1 notification message, got %d", len(notifications.Messages))
	}

	if len(audit.Entries) != 1 {
		t.Fatalf("expected 1 audit entry, got %d", len(audit.Entries))
	}
}

func TestPlaceOrderRejectsInvalidTotal(t *testing.T) {
	service := NewPlacementService(NewEventBus())

	if _, err := service.PlaceOrder("ord-5002", "cust-200", 0); err == nil {
		t.Fatal("expected invalid total to fail")
	}
}

File: examples/typescript/order-domain-event/order/service.test.ts
import { AuditHandler, EventBus, NotificationHandler } from "./event";
import { PlacementService } from "./service";

function assert(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

function assertThrows(run: () => void, expectedMessage: string): void {
  try {
    run();
    throw new Error("expected function to throw");
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    assert(message.includes(expectedMessage), `expected error to include "${expectedMessage}", got "${message}"`);
  }
}

const bus = new EventBus();
const notifications = new NotificationHandler();
const audit = new AuditHandler();

bus.subscribe((event) => notifications.handleOrderPlaced(event));
bus.subscribe((event) => audit.handleOrderPlaced(event));

const service = new PlacementService(bus);
const placedOrder = service.placeOrder({
  orderId: "ord-ts-5001",
  customerId: "cust-ts-100",
  totalCents: 15900
});

assert(placedOrder.id === "ord-ts-5001", "expected placed order id");
assert(notifications.messages.length === 1, "expected one notification message");
assert(audit.entries.length === 1, "expected one audit entry");

assertThrows(
  () =>
    service.placeOrder({
      orderId: "ord-ts-5002",
      customerId: "cust-ts-200",
      totalCents: 0
    }),
  "total must be positive"
);

console.log("order-domain-event tests passed");
